File: bin/native/bend2/effs/udp_poll.js
// UDP
// ===

function udp_poll(socket, max) {
  const sys = io_sys();
  const fd = socket;
  const b = new Uint8Array(Math.max(Number(max), 1));
  const peer = new Uint8Array(16);
  const len = new Uint32Array([16]);
  const got = sys.recvfrom(fd, sys.ptr(b), Number(max), 0, sys.ptr(peer),
    sys.ptr(len));
  const n = Number(got);
  if (n < 0 && sys.errno() === (sys.mac ? 35 : 11)) {
    return io_tup(socket, io_done({ $: "None" }));
  }
  if (n < 0) {
    return io_tup(socket, io_fail(sys.errno()));
  }
  const host = peer[4] + "." + peer[5] + "." + peer[6] + "." + peer[7];
  const port = (peer[2] << 8) | peer[3];
  return io_tup(socket, io_done({ $: "Some",
    value: io_tup(host, port, io_text(b, n)) }));
}

File: bin/native/bend2/effs/udp_poll.c
// UDP
// ===

Term udp_poll_run(Env e, Term* f, IoWork* w) {
  struct sockaddr_in at = { 0 };
  socklen_t alen = sizeof(at);
  char      host[16];
  int       fd   = (int)io_hand_v(f[0]);
  u32       max  = f[1] < INT32_MAX ? (u32)f[1] : INT32_MAX;
  char*     data = io_mem(malloc(max + 1));
  ssize_t   n    = recvfrom(fd, data, max, 0, (struct sockaddr*)&at, &alen);
  u32       code = n < 0 ? (u32)errno : 0;
  Term      r;
  if (code == EAGAIN) {
    r = io_done(e, term_pak(CID_NONE, 0));
  } else if (code != 0) {
    r = io_fail(e, code, NULL);
  } else {
    inet_ntop(AF_INET, &at.sin_addr, host, 16);
    r = io_done(e, io_box(e, CID_SOME, io_tup(e, io_str(e, host, strlen(host)),
      io_tup(e, ntohs(at.sin_port), io_str(e, data, (u64)n)))));
  }
  free(data);
  return io_tup(e, f[0], r);
}

static void __attribute__((constructor)) udp_poll_use(void) {
  io_eff(CID_UDP_POLL, udp_poll_run, 0);
}
